src/client/statenav.c

```diff
@@ -346,10 +350,108 @@ void receiveButtonClick(char* value)
       int surrender = strtol(value + strlen("surrender "), NULL, 10);
       sendOrder(SURRENDERORDER, UINT_MAX, surrender);
    }
+   else if (!strncmp(value, "scoreboard", strlen("scoreboard")))
+   {
+      showScoreboard();
+   }
 
    free(value);
 }
 
+void refreshScoreboardIfNecessary()
+{
+   int scoreboardIsVisible = EM_ASM_INT(
+      {
+         return document.getElementById('scoreboard') != null ? 1 : 0;
+      });
+
+   if (scoreboardIsVisible)
+   {
+      showScoreboard();
+   }
+}
+
+void showScoreboard()
+{
+   EM_ASM(
+      {
+         if (document.getElementById('scoreboard'))
+         {
+            document.getElementById('scoreboard').remove();
+         }
+
+         const controlledBy = Module.HEAPU32.subarray($1/4, $1/4 + $0);
+         const playerCount = $2;
+         const playerNames = Module.HEAPU32.subarray($3/4, $3/4 + $2);
+         const playerColors = Module.HEAPU32.subarray($4/4, $4/4 + $2);
+         let scoreboard = [];
+
+         for (let player = 0; player < playerCount; ++player)
+         {
+            let name = UTF8ToString(playerNames[player]);
+            let color = UTF8ToString(playerColors[player]);
+
+            scoreboard[player] = {};
+            scoreboard[player].name = name;
+            scoreboard[player].color = color;
+            scoreboard[player].nodes = 0;
+         }
+
+         let neutralCount = 0;
+         for (let node = 0; node < $0; ++node)
+         {
+            if (controlledBy[node] != 4294967295)
+            {
+               scoreboard[controlledBy[node]].nodes += 1
+            }
+            else
+            {
+               neutralCount++;
+            }
+         }
+
+         scoreboard.sort((a, b) => b.nodes - a.nodes);
+
+         let scoreHtml = `
+            <div id="scoreboard">
+               <p>Total nodes: ${$0}<br>Neutral: ${neutralCount}</p>
+               <ul>
+         `;
+
+         for (const player of scoreboard)
```

Comment:
A bit subjective. But I think this section would be more readable using a string base apprach.
```suggestion
         const listHtml = scoreboard.map(player => (`
                <li><span style="color: ${player.color};">&#x2588;  </span><span>${player.name}: ${player.nodes}</span></li>
         `)).join("")
```
Also it is less imperative which is also a subjective benefit :P

Reply:
This is very nice and short! But the problem is that player.color and player.name are untrusted user input, so this would allow arbitrary HTML. E.g., if a player chooses the name `<img src="https://i.imgur.com/HLHBnl9.jpeg">`, we'd end up with this:

![2021-06-07_13-33](https://user-images.githubusercontent.com/906563/121009982-19182780-c795-11eb-9c2a-1b039ec3bab9.png)

Hence the more laborious approach :)